### PR TITLE
262 saas connector pagination strategies

### LINKED ISSUE #262: [SaaS Connector] Pagination strategies
Add the ability for each SaaS connector endpoint to be able to specify a pagination strategy. These three strategies should be implemented as part of this work.

**offset** - Takes a `SaasRequestParams` tuple and updates the query param with name of `incremental_param` and updates the value according to `initial_value`, `increment_by` and `page_limit`
```
pagination:
  strategy: offset
  configuration:
    incremental_param: page # which query, path, or body param to update with the new value
    increment_by: 1
    limit: 100
```
In this case `page` is the incremental param, so we must also enforce that the `page` query param has a `default_value` to start the pagination with (under request_params)

**link** - Get a link in the headers or response and follow it to retrieve the next page of data. Replace the URL in the `SaaSRequestParams` with the new URL. Verify host name matches the configured host name for the given connector. The `path` parameter works in the same way as the data path does in the `unwrap` post-processor strategy.
```
pagination:
  strategy: link
  configuration
    source: headers or response
    path: pages.next.starting_after
```

**cursor** - Get a cursor value and pass it in as a query param to get the next set of results. This will take the read the `id` field of the post-processed object and use it as the `start_after` query param
```
pagination:
  strategy: cursor
  configuration:
    cursor_param: start_after
    value: id
```

**Additional Requirements**
- The data returned from each "page" should run through the post-processors before retrieving the next page to reduce memory usage.
- Pagination should stop once no results are returned, or the page limit is hit (for the offset paging strategy).

### PR DESCRIPTION
**Pull request by galvana**
# Purpose
To add pagination strategies for SaaS connectors

# Changes
- Added offset, link, and cursor pagination strategies
- Added pagination to `retrieve_data` function in SaaS Connector
- Updated documentation with pagination strategies
- Updated Postman collection
- Changed saas_config spec to define the unwrap filter as a standalone `data_path` on the request to reduce verbosity and to prevent pagination strategies (which need the data_path) from depending on postprocessor configurations.

# Checklist

- [ ] Applicable documentation updated (guides, quickstart, postman collections, tutorial, fidesdemo, [database diagram](https://github.com/ethyca/fidesops/blob/main/docs/fidesops/docs/development/update_erd_diagram.md))
- [ ] Good unit test/integration test coverage
- [ ] This PR contains a DB migration. If checked, the reviewer should confirm with the author that the [down_revision correctly references the previous migration](https://ethyca.github.io/fidesops/development/contributing_details/#alembic-migrations) before merging
- [ ] The `Run Unsafe PR Checks` label has been applied, and checks have passed, if this PR touches any external services

# Ticket

Closes #262

### CODE REVIEW

**Comment by eastandwestwind** [comment]
I have some concerns around whether we can always assume that our data path for pagination will be the exact same as that needed for unwrap post-processing. For example, if pagination data is within a diff data path than the actual identity data we need to extract?

Further, what if we have a use case where we need to unwrap, filter, then unwrap a 2nd time (using a diff data path)? What do you think @galvana ?

**Comment by galvana** [comment]
> I have some concerns around whether we can always assume that our data path for pagination will be the exact same as that needed for unwrap post-processing. For example, if pagination data is within a diff data path than the actual identity data we need to extract?
> 
> Further, what if we have a use case where we need to unwrap with data path `a`, filter, then unwrap again, using data path `b`? What do you think @galvana ?

Hey @eastandwestwind these are valid concerns, here are my thoughts on these two points:

1) With regards to pagination, the data_path is just used to see if there is any data at all in that location (before filtering or other post-processing). This is how we determine if we should generate a subsequent request for more data. It could be the case that a specific endpoint call doesn't result in any valid data after postprocessing but if the raw response still had data at the `data_path` location, we want to keep paginating until nothing is returned (an empty list).

2) The unwrap filter is still available for more complex scenarios like this one. We would just have to unwrap based on where the data is now after the default `data_path` unwrap and any filter postprocessors.

**Comment by eastandwestwind** [comment]
To elaborate, **for #1**,  if API response:

```
 [
    {"email": {"personal": "me@email.com"}},
    {"email": {"personal": "someone@email.com"}}
  ]
```

And we first needed a filter processor, then an unwrap with `data_path` = `email`, to retrieve the resulting data: `{"personal": "me@email.com"}`

In this scenario, if we also needed pagination, what would our `data_path` be? I don't think  [pydash.get(response.json(), data_path)](https://github.com/ethyca/fidesops/pull/286/files#diff-9743c52786ccb3dc91d2ea2619e009089e590880dd19e4d6797355615fd5e779R34) supports data_paths that are within arrays? Is this a valid scenario?



**For #2:** if we have the following API response:

```
{
"results":
  [
    {"info": {"email": "who@email.com", "preferences": {}}},
    {"info": {"other@email.com", "preferences": {}}}
  ]
}
```

Where we first implement an unwrap processor such that `data_path` = `results`, then a filter to get the relevant data, then another unwrap such that `data_path` = `info`, to retrieve the resulting data: `{"email": "who@email.com", "preferences": {}}`.  Do you think we'd ever run into this situation where we'd need a diff data_path per unwrap processor?

**Comment by galvana** [comment]
Thanks for the examples, it helped uncover the scenario where the response data can be found at the root level of the response and does not need a data_path. If we make the data_path optional for these scenarios, then this is how we can handle the scenarios you specified.

### Scenario 1:
```
[
  {"email": {"personal": "me@email.com"}},
  {"email": {"personal": "someone@email.com"}}
]
```

In this case we can omit the data_path from the request and use this configuration:
```
postprocessors:
  - strategy: unwrap
    configuration
      data_path: email
  - strategy: filter
     configuration:
       field: personal
       value:
         identity: email
pagination:
  strategy: offset
  configuration:
    incremental_param: page
    increment_by: 1
    limit: 10
```
The expected results from this would be:
```
[
  {"personal": "me@email.com"}
]
```
And the pagination strategy would see 2 results in the list (because of the empty data_path) and would continue to the next page.

### Scenario 2:
```
{
"results":
  [
    {"info": {"email": "who@email.com", "preferences": {}}},
    {"info": {"email": "other@email.com", "preferences": {}}}
  ]
}
```
For this scenario we could do this:
```
data_path: results
postprocessors:
  - strategy: unwrap
    configuration
      data_path: info
  - strategy: filter
     configuration:
       field: personal
       value:
         identity: email
pagination:
  strategy: offset
  configuration:
    incremental_param: page
    increment_by: 1
    limit: 10
```
The result would be:
```
{"email": "who@email.com", "preferences": {}}
```
And the pagination strategy would also see the two results after the implicit `data_path` unwrap and continue paging.

**Comment by eastandwestwind** [comment]
This helps a lot @galvana ! So to summarize, we'll keep `data_path` as optional both on the postprocessor, AND at the more generic endpoint level. 

So, a postprocessor will first look within the postprocessor config, and, if not found there, will look at the endpoint level. Is that right?

**Comment by galvana** [comment]
> This helps a lot @galvana ! So to summarize, we'll keep `data_path` as optional both on the postprocessor, AND at the more generic endpoint level.
> 
> So, a postprocessor will first look within the postprocessor config, and, if not found there, will look at the endpoint level. Is that right?

To the first point, yes, and empty data_path can be the way we denote "unwrapping" an array at both the endpoint and postprocessor level. For the second point, the postprocessor will only look in its config for the data_path. The only way the endpoint data_path will impact postprocessors is that it will affect what gets passed into the postprocssors (the data found at data_path instead of the raw response). I'll at some test cases for the scenarios we discussed here.

File: src/fidesops/service/processors/post_processor_strategy/post_processor_strategy_unwrap.py [hunk 1]
```diff
@@ -43,23 +45,32 @@ def get_strategy_name(self) -> str:
...
+        if isinstance(data, dict):
+            unwrapped = pydash.get(data, self.data_path)
+
+        if isinstance(data, list):
+            unwrapped = []
+            for item in data:
+                unwrapped.append(pydash.get(item, self.data_path))
+            # flatten the list to account for the event where the output of unwrapped
+            # is a list of lists
+            unwrapped = pydash.flatten(unwrapped)
```
[the remark this answers is not in the document]
Comment (galvana) [reply src/fidesops/service/processors/post_processor_strategy/post_processor_strategy_unwrap.py:67]: Added!

File: src/fidesops/service/pagination/pagination_strategy_cursor.py [hunk 2]
```diff
@@ -28,11 +28,12 @@ def get_next_request(
...
         """Build request for next page of data"""
 
-        # get the last object in the array specified by data_path
-        # and read the cursor value
+        # get the last object in the array and read the cursor value
         cursor = None
-        object_list = pydash.get(response.json(), data_path)
-        if object_list:
+        object_list = (
+            pydash.get(response.json(), data_path) if data_path else response.json()
```
[the remark this answers is not in the document]
Comment (eastandwestwind) [reply src/fidesops/service/pagination/pagination_strategy_cursor.py:34]: this makes sense to me, thanks for the detailed explanation!

**Review by eastandwestwind — APPROVED** [review]
Looks great, thanks for all the hard work addressing the additional use cases @galvana !

**Pull request merged**